core/domain/services/checkout/fees/ExtraTxnFeesHandler.php

```diff
@@ -0,0 +1,103 @@
+<?php
+
+namespace EventEspresso\core\domain\services\checkout\fees;
+
+use DomainException;
+use EE_Error;
+use EE_Registration;
+use EE_Transaction;
+use ReflectionException;
+
+/**
+ * Class ExtraTxnFeesHandler
+ *
+ * @author  Brent Christensen
+ * @package EventEspresso\core\domain\services\checkout\extra_fees
+ * @since   $VID:$
+ */
+class ExtraTxnFeesHandler
+{
+    /**
+     * @param EE_Transaction $transaction
+     * @param EE_Registration[] $registrations
+     * @return void
+     * @throws DomainException
+     * @throws EE_Error
+     * @throws ReflectionException
+     */
+    public function applyExtraFeesToRegistrants(EE_Transaction $transaction, array $registrations)
+    {
+        $extra_fees = $this->calculateExtraFees($transaction, $registrations);
+        // if there's money owing after all the individual registration fees have been subtracted
+        if ($extra_fees > 0) {
+            // then first decide whether to spread extra fees across all registrations,
+            // or assign everything to the primary registrant
+            $distribution_strategy = $this->getExtraTxnFeesDistributionStrategy($transaction);
+            $distribution_strategy->applyExtraFeesToRegistrants($extra_fees, $registrations);
+        }
+    }
+
+
+    /**
+     * @param EE_Transaction    $transaction
+     * @param EE_Registration[] $registrations
+     * @return float
+     * @throws EE_Error
+     * @throws ReflectionException
+     */
+    private function calculateExtraFees(EE_Transaction $transaction, array $registrations)
+    {
+        // we're going to calculate if there are any extra fees on the TXN, so start with the full amount
+        $extra_fees = $transaction->total();
+        // then subtract the total for each registration from what's owing for the entire txn
+        foreach ($registrations as $registration) {
+            if ($registration instanceof EE_Registration) {
+                $extra_fees -= $registration->final_price();
+            }
+        }
+        return $extra_fees;
```

Comment:
Is this kind of business logic documented anywhere?

Reply:
no, unfortunately we don't have any of our business logic documented anywhere (yet), but this is a bit of a new situation since it only arises if you add that extra functionality from the gist. There is some similar logic that distributes extra fees across the line items but that was never applied directly to the registration payments.